segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java

```diff
@@ -0,0 +1,210 @@
+            val keyIterator = tableExtension.keyIterator(getMetadataSegmentName(
+                    containerEntry.getKey()), args).get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+
+            // Store the segments in a set
+            Set<String> metadataSegments = new HashSet<>();
+            keyIterator.forEachRemaining(k ->
+                    metadataSegments.addAll(k.getEntries().stream()
+                            .map(entry -> entry.getKey().toString())
+                            .collect(Collectors.toSet())), executorService).get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+            metadataSegmentsMap.put(containerEntry.getKey(), metadataSegments);
+        }
+        return metadataSegmentsMap;
+    }
+
+    /**
+     * This method takes a {@link DebugStreamSegmentContainer} instance and a {@link SegmentProperties} object as arguments
+     * and takes one of the following actions:
+     * 1. If the segment is present in the container metadata and its length or sealed status or both doesn't match with the
+     * given {@link SegmentProperties}, then it is deleted from there and registered using the properties from the given
+     * {@link SegmentProperties} instance.
+     * 2. If the segment is absent in the container metadata, then it is registered using the properties from the given
+     * {@link SegmentProperties}.
+     * @param container         A {@link DebugStreamSegmentContainer} instance for registering the given segment and checking
+     *                          its existence in the container metadata.
+     * @param storageSegment    A {@link SegmentProperties} instance which has properties of the segment to be registered.
+     * @return                  CompletableFuture which when completed will have the segment registered on to the container
+     *                          metadata.
+     */
+    private static CompletableFuture<Void> recoverSegment(DebugStreamSegmentContainer container, SegmentProperties storageSegment) {
+        Preconditions.checkNotNull(container);
+        Preconditions.checkNotNull(storageSegment);
+        long segmentLength = storageSegment.getLength();
+        boolean isSealed = storageSegment.isSealed();
+        String segmentName = storageSegment.getName();
+
+        log.info("Registering: {}, {}, {}.", segmentName, segmentLength, isSealed);
+        return Futures.exceptionallyComposeExpecting(
+                container.getStreamSegmentInfo(storageSegment.getName(), TIMEOUT)
+                        .thenAccept(e -> {
+                            if (segmentLength != e.getLength() || isSealed != e.isSealed()) {
+                                container.metadataStore.deleteSegment(segmentName, TIMEOUT)
+                                        .thenAccept(x -> container.registerSegment(segmentName, segmentLength, isSealed));
+                            }
+                        }), ex -> Exceptions.unwrap(ex) instanceof StreamSegmentNotExistsException,
```

Comment:
I am pretty sure that `Futures.exceptionallyComposeExpecting` unwraps it for you. Please check, and if so, simplify this line.

Reply:
Yeah. It does. Thanks.